### PR TITLE
[SPARK-19915][SQL] Exclude cartesian product candidates to reduce the search space

### PR DESCRIPTION
**Pull request by wzhfy**
## What changes were proposed in this pull request?

We have some concerns about removing size in the cost model [in the previous pr](https://github.com/apache/spark/pull/17240). It's a tradeoff between code structure and algorithm completeness. I tend to keep the size and thus create this new pr without changing cost model.

What this pr does:
1. It deals with consecutive inner joinable items. During reordering, it excludes cartesian products in reordering procedure. 
Given A Join B on A.key = B.key Join C on B.value = C.value:
level 0: {A}, {B}, {C}
level 1: {A, B}, ~~{A, C}~~, {B, C}
level 2: {A, B, C}
When the number of items increases, this significantly reduces the search space and memory overhead of memo. Otherwise every combination of items will exist in the memo.

2. This pr also includes a bug fix: if an item (basic reordering unit) is a project(_, child), current solution will lost the project.

## How was this patch tested?

Added test cases.

### CODE REVIEW

**Comment by wzhfy** [comment]
please review @hvanhovell @cloud-fan @nsyca

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 1]
```diff
@@ -19,31 +19,31 @@ package org.apache.spark.sql.catalyst.optimizer
...
  */
-case class CostBasedJoinReorder(conf: CatalystConf) extends Rule[LogicalPlan] with PredicateHelper {
+case class CostBasedJoinReorder(conf: SQLConf) extends Rule[LogicalPlan] with PredicateHelper {
   def apply(plan: LogicalPlan): LogicalPlan = {
     if (!conf.cboEnabled || !conf.joinReorderEnabled) {
       plan
     } else {
-      val result = plan transform {
+      val result = plan transformDown {
         case p @ Project(projectList, j @ Join(_, _, _: InnerLike, _)) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: shall we also check if `projectList` are all attributes here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 2]
```diff
@@ -128,38 +128,43 @@ case class CostBasedJoinReorder(conf: CatalystConf) extends Rule[LogicalPlan] wi
...
-    val plansLastLevel = foundPlans(items.length - 1)
-    if (plansLastLevel.isEmpty) {
-      // Failed to find a plan, fall back to the original plan
-      None
-    } else {
-      // There must be only one plan at the last level, which contains all items.
-      assert(plansLastLevel.size == 1 && plansLastLevel.head._1.size == items.length)
-      Some(plansLastLevel.head._2.plan)
+    // Find the best plan
+    assert(foundPlans.last.size <= 1)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: can you answer this question: https://github.com/apache/spark/pull/17240#discussion_r105822819

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala (hunk 1 above)
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: Yea I think so.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala (hunk 2 above)
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: hmm you have a good point here. If we have several disconnect item sets, e.g. {AB} and {CD}, or a more complex case: {ABCD}, {EFG}, {LM}... These cases need to be dealt with.

Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: Seems excluding CP bring some complexity here.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: how about
```
while (foundPlans.size < items.length && foundPlans.last.size > 0)
```
When we end the while loop, either we have reached the level n, or the current level has 0 entries. Then we pick the last level which has non-zero entries, and pick the best entry from this level, and construct the final join plan.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -696,6 +696,13 @@ object SQLConf {
       .intConf
       .createWithDefault(12)
 
+  val JOIN_REORDER_CARD_WEIGHT =
+    buildConf("spark.sql.cbo.joinReorder.card.weight")
+      .doc("The weight of cardinality (number of rows) for plan cost comparison in join reorder: " +
+        "rows * weight + size * (1 - weight).")
+      .doubleConf
+      .createWithDefault(0.7)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: What is boundary of this? adding `check`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala (hunk 2 above)
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: We still cannot deal with the case where several disconnected groups of items exist in the memo. 
Join relations may exist within these groups. E.g.
```
level 3: {ABCD}
level 2: {EFG}, {HIJ}, {ABC}...
level 1: {KL}, {MN}, {AB}, {BC}...
level 0: ...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 3 above)
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Yes, checking should be added, thanks!

**Comment by wzhfy** [comment]
Co-existing cross joins (join without a condition) and inner joins make the reordering procedure cumbersome. On one hand, putting cross join candidates into memo is not good in terms of search performance and memory consumption. On the other hand, adding cross joins after inner join exploration also has problems with multiple unjoinable groups, just putting them at the end of the plan is not right.

So I decide to do reordering only for consecutive inner joins, which are separated by other plans (including cross joins). For bushy trees, this means we will reorder joins within each inner joinable "groups". For left deep trees, the `ReorderJoin` rule puts cross joins to the end, so all the previous inner joins will be reordered. Since join reorder is a complicated problem, we may need several levels of optimization, we can consider `ReorderJoin` as a heuristic algorithm.

For common cases when there's no cross join, this algorithm behaves just like before.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 4]
```diff
@@ -128,38 +131,34 @@ case class CostBasedJoinReorder(conf: CatalystConf) extends Rule[LogicalPlan] wi
...
+      conf: SQLConf,
       items: Seq[LogicalPlan],
       conditions: Set[Expression],
-      topOutput: AttributeSet): Option[LogicalPlan] = {
+      topOutput: AttributeSet): LogicalPlan = {
 
     // Level i maintains all found plans for i + 1 items.
     // Create the initial plans: each plan is a single item with zero cost.
-    val itemIndex = items.zipWithIndex
+    val itemIndex = items.zipWithIndex.map(_.swap).toMap
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: looks like an unnecessary change now

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 5]
```diff
@@ -185,11 +184,14 @@ object JoinReorderDP extends PredicateHelper {
           // Should not join two overlapping item sets.
           if (oneSidePlan.itemIds.intersect(otherSidePlan.itemIds).isEmpty) {
             val joinPlan = buildJoin(oneSidePlan, otherSidePlan, conf, conditions, topOutput)
-            // Check if it's the first plan for the item set, or it's a better plan than
-            // the existing one due to lower cost.
-            val existingPlan = nextLevel.get(joinPlan.itemIds)
-            if (existingPlan.isEmpty || joinPlan.cost.lessThan(existingPlan.get.cost)) {
-              nextLevel.update(joinPlan.itemIds, joinPlan)
+            if (joinPlan.isDefined) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: when will this condition be false?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 6]
```diff
@@ -272,26 +256,39 @@ object JoinReorderDP extends PredicateHelper {
    * @param itemIds Set of item ids participating in this partial plan.
    * @param plan The plan tree with the lowest cost for these items found so far.
    * @param joinConds Join conditions included in the plan.
-   * @param cost The cost of this plan is the sum of costs of all intermediate joins.
+   * @param planCost The cost of this plan tree is the sum of costs of all intermediate joins.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: I think `cost` is good enough, why rename it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 7]
```diff
@@ -203,64 +205,46 @@ object JoinReorderDP extends PredicateHelper {
...
-    val collectedJoinConds = joinConds ++ oneJoinPlan.joinConds ++ otherJoinPlan.joinConds
-    val remainingConds = conditions -- collectedJoinConds
-    val neededAttr = AttributeSet(remainingConds.flatMap(_.references)) ++ topOutput
-    val neededFromNewJoin = newJoin.outputSet.filter(neededAttr.contains)
-    val newPlan =
-      if ((newJoin.outputSet -- neededFromNewJoin).nonEmpty) {
-        Project(neededFromNewJoin.toSeq, newJoin)
+    if (joinConds.isEmpty) {
+      // Cartesian product is very expensive, so we exclude them from candidate plans.
+      // This also significantly reduces the search space.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: great! now we can safely apply this optimization :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 8]
```diff
@@ -710,6 +710,14 @@ object SQLConf {
       .intConf
       .createWithDefault(12)
 
+  val JOIN_REORDER_CARD_WEIGHT =
+    buildConf("spark.sql.cbo.joinReorder.card.weight")
+      .doc("The weight of cardinality (number of rows) for plan cost comparison in join reorder: " +
+        "rows * weight + size * (1 - weight).")
+      .doubleConf
+      .checkValue(weight => weight >= 0 && weight <= 1, "The weight value must be in [0, 1].")
+      .createWithDefault(0.7)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: it is useful to expose this config? I think most of the users will just disable join reordering if they have problems.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinReorderSuite.scala [hunk 9]
```diff
@@ -187,6 +220,8 @@ class JoinReorderSuite extends PlanTest with StatsEstimationTestBase {
       case (j1: Join, j2: Join) =>
         (sameJoinPlan(j1.left, j2.left) && sameJoinPlan(j1.right, j2.right)) ||
           (sameJoinPlan(j1.left, j2.right) && sameJoinPlan(j1.right, j2.left))
+      case _ if plan1.children.nonEmpty && plan2.children.nonEmpty =>
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinReorderSuite.scala]: when will we hit this branch?

**Comment by cloud-fan** [comment]
LGTM except some minor comments

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 8 above)
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: `0.7` is just an empirical value, I think it's better to be able to tune it. Maybe mark it internal?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinReorderSuite.scala (hunk 9 above)
Comment (wzhfy) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/JoinReorderSuite.scala]: If they are two projects, we should recurse into it's child. And we may have some tests with other operators in the future, so I didn't only match project here.

**Comment by cloud-fan** [comment]
retest this please

**Comment by ioana-delaney** [comment]
@wzhfy Some thoughts on how to solve the Cartesian problem as part of the join enumeration algorithm is to apply a similar strategy to the one that we discuss for star-plans. You keep track of "connected" tables and "unconnected" tables. During join enumeration, mixed combinations are pruned until the plan for the “connected” set of tables was built. Then, we add tables from the "unconnected" set - maybe only as left-deep trees (i.e. the size of the inner is one). Also, knowing that a set of tables are connected through join conditions, will allow further plan pruning based on the presence of join predicates. Integration with star-join, and probably other heuristics, would require to introduce some filtering/pruning strategies on top of the search engine. Just some thoughts…

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 10]
```diff
@@ -19,31 +19,33 @@ package org.apache.spark.sql.catalyst.optimizer
...
     if (!conf.cboEnabled || !conf.joinReorderEnabled) {
       plan
     } else {
-      val result = plan transform {
-        case p @ Project(projectList, j @ Join(_, _, _: InnerLike, _)) =>
-          reorder(p, p.outputSet)
-        case j @ Join(_, _, _: InnerLike, _) =>
+      val result = plan transformDown {
+        // Start reordering with a joinable item, which is an InnerLike join with conditions.
+        case j @ Join(_, _, _: InnerLike, Some(cond)) =>
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: I have a very general question about how to exclude cartesian product. See my old PR: #16762

Having a join condition does not mean it is not a Cartesian product. : )

Only having non-equal predicates in join conditions for an inner join. e.g., tab1.a < tab2.b

**Comment by wzhfy** [comment]
@ioana-delaney How can we get disconnected parts before reordering? They can not only be leaf tables (this case is easier to deal with), but also subplans (tables can be joined internally but among subplans there's no join condition). We need to consider both cases and they may co-exist. Also, some disconnection can be eliminated by reordering.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala [hunk 11]
```diff
@@ -19,31 +19,33 @@ package org.apache.spark.sql.catalyst.optimizer
...
           reorder(j, j.outputSet)
+        case p @ Project(projectList, Join(_, _, _: InnerLike, Some(cond)))
+          if projectList.forall(_.isInstanceOf[Attribute]) =>
+          reorder(p, p.outputSet)
       }
       // After reordering is finished, convert OrderedJoin back to Join
-      result transform {
+      result transformDown {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala]: Is the change needed? The order must be top-down?

**Comment by ioana-delaney** [comment]
@wzhfy Given a set of input plans (either base table access or plans over derived/complex plans), one can build a graph based on the join conditions among the plans. I think join enumeration should be driven by the presence of join conditions between two plans. Following this approach, I think we can detect items that are not connected and plan them towards the end.

**Comment by gatorsmile** [comment]
Given {A, B, C} with the join conditions {A.a = B.b, B.b > C.c}. `{A, C}` should be pruned.

**Comment by nsyca** [comment]
@gatorsmile An equality join in most cases has a better filtering than an inequality join. This can be used heuristically. However, this is not always true. An equality join can be a lookup join from a fact table to a dimension table that it does not reduce anything but an inequality join such as a range join over time period.

**Comment by gatorsmile** [comment]
My example is not related to inequality join or equi join.

**Comment by ioana-delaney** [comment]
@gatorsmile Your example is correct. 

Given A J1 B J2 C:
•	level 0: (A), (B), (C)
•	level 1: {A, B}, ~{A, C}~, {B, C}
•	level 3: {A, B, C}

Given A J1 B J2 C cross join D 
•	level 0: (A), (B), (C), (D)
•	level 1: {A, B}, ~{A, C}~, ~{A, D}~, {B, C}, ~{B, D}~
•	level 3: {A, B, C}, ~{A, B, D}~, ~{A, C, D}~, ~{B, C, D}~
•	level 4: {A, B, C, D}

The presence of join conditions between the outer-inner sets are good heuristics to be used during join enumeration.

**Comment by nsyca** [comment]
Right. I misread it. if there is no join predicate between a table and any cluster of tables, we should not consider that table in the join enumeration at all. We can simply push that table to be the last join.

**Comment by wzhfy** [comment]
@ioana-delaney Thanks for the replies.

> Given A J1 B J2 C:   - case 1
•	level 0: (A), (B), (C)
•	level 1: {A, B}, ~~{A, C}~~, {B, C}
•	level 3: {A, B, C}
Given A J1 B J2 C cross join D  - case 2
•	level 0: (A), (B), (C), (D)
•	level 1: {A, B}, ~~{A, C}, {A, D}~~, {B, C}, ~~{B, D}~~
•	level 3: {A, B, C}, ~~{A, B, D}, {A, C, D}, {B, C, D}~~
•	level 4: {A, B, C, D}

Currently the algorithm can prune cartesian products in case 1. For case 2, now it is based on the output of `ReorderJoin`, cartesian products are put at the end. So we have the same effect on the final plan.

In the future, we can merge these rules together.
